File: seminar-09/dsu/python/dsu.py
class DSU:
    def __init__(self, size):
        self.parent = [i for i in range(size)]
        self.size = [1 for _ in range(size)]

    def get(self, x):
        if self.parent[x] == x:
            return x
        leader = self.get(self.parent[x])
        self.parent[x] = leader
        return leader

    def is_same(self, a, b):
        return self.get(a) == self.get(b)

    def union(self, a, b):
        a = self.get(a)
        b = self.get(b)
        if a == b:
            return False

        if self.size[a] < self.size[b]:
            a, b = b, a

        self.parent[b] = a
        self.size[a] += self.size[b]
        return True

    def component_size(self, x):
        return self.size[self.get(x)]

File: seminar-09/dsu/java/Dsu.java
public class Dsu {

    int[] parent;
    int[] size;

    public Dsu(int n) {
        parent = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    public int get(int x) {
        if (parent[x] == x) return x;
        return parent[x] = get(parent[x]);
    }

    public boolean isSame(int a, int b) {
        return get(a) == get(b);
    }

    public boolean union(int a, int b) {
        a = get(a);
        b = get(b);
        if (a == b) return false;

        if (size[a] < size[b]) {
            int tmp = a;
            a = b;
            b = tmp;
        }

        parent[b] = a;
        size[a] += size[b];
        return true;
    }

    public int size(int x) {
        return size[get(x)];
    }
}
